src/moscot/utils/tagged_array.py

```diff
@@ -159,3 +172,205 @@ def is_kernel(self) -> bool:
+        conditions_key: Optional[str],
+    ) -> Tuple[bool, bool, bool]:
+        if (xy_attr is None and xy_key is not None) or (xy_attr is not None and xy_key is None):
+            raise ValueError(r"Either both `xy_attr` and `xy_key` must be `None` or none of them.")
+        if (xx_attr is None and xx_key is not None) or (xx_attr is not None and xx_key is None):
+            raise ValueError(r"Either both `xy_attr` and `xy_key` must be `None` or none of them.")
+        if (conditions_attr is None and conditions_key is not None) or (
+            conditions_attr is not None and conditions_key is None
+        ):
+            raise ValueError(r"Either both `conditions_attr` and `conditions_key` must be `None` or none of them.")
+        return xy_attr is not None, xx_attr is not None, conditions_attr is not None
+
+    @classmethod
+    def from_adata(
+        cls,
+        adata: AnnData,
+        a: ArrayLike,
+        b: ArrayLike,
+        xy_attr: Literal["X", "obsp", "obsm", "layers", "uns"] = None,
+        xy_key: Optional[str] = None,
+        xy_cost: CostFn_t = "sq_euclidean",
+        xx_attr: Literal["X", "obsp", "obsm", "layers", "uns"] = None,
+        xx_key: Optional[str] = None,
+        xx_cost: CostFn_t = "sq_euclidean",
+        conditions_attr: Optional[Literal["obs", "var", "obsm", "varm", "layers", "uns"]] = None,
+        conditions_key: Optional[str] = None,
+        backend: Literal["ott"] = "ott",
+        **kwargs: Any,
+    ) -> "DistributionContainer":
+        """Create distribution container from :class:`~anndata.AnnData`.
+
+        .. warning::
+            Sparse arrays will be always densified.
+
+        Parameters
+        ----------
+        adata
+            Annotated data object.
+        a
+            Marginals when used as source distribution.
+        b
+            Marginals when used as target distribution.
+        xy_attr
+            Attribute of :paramref:`adata` containing the data for the shared space.
+        xy_key
+            Key of :paramref:`xy_attr` containing the data for the shared space.
+        xy_cost
+            Cost function when in the shared space.
+        xx_attr
+            Attribute of :paramref:`adata` containing the data for the incomparable space.
+        xx_key
+            Key of :paramref:`xx_attr` containing the data for the incomparable space.
+        xx_cost
+            Cost function in the incomparable space.
+        conditions_attr
+            Attribute of :paramref:`adata` containing the conditions.
+        conditions_key
+            Key of :paramref:`conditions_attr` containing the conditions.
+        backend
+            Backend to use.
+        kwargs
+            Keyword arguments to pass to the cost functions.
+
+        Returns
+        -------
+        The distribution container.
+        """
+        contains_linear, contains_quadratic, contains_condition = cls._verify_input(
```

Comment:
Would consider adding these as properties, might be useful when using this container.